api-put-bucket.go

```diff
@@ -176,10 +176,17 @@ func (c Client) SetBucketPolicy(bucketName string, objectPrefix string, bucketPo
 	}
 	statements = append(statements, generatedStatements...)
 
-	// No change in the statements indicates an attempt of setting 'none' on a prefix
-	// which doesn't have a pre-existing policy.
+	// No change in the statements indicates either an attempt of setting 'none'
+	// on a prefix which doesn't have a pre-existing policy, or setting a policy
+	// on a prefix which already has the same policy.
 	if reflect.DeepEqual(policy.Statements, statements) {
-		return ErrNoSuchBucketPolicy(fmt.Sprintf("No policy exists on %s/%s", bucketName, objectPrefix))
+		// If policy being set is 'none' return an error, otherwise return nil to
+		// prevent the unnecessary request from being sent
+		if bucketPolicy == BucketPolicyNone {
+			return ErrNoSuchBucketPolicy(fmt.Sprintf("No policy exists on %s/%s", bucketName, objectPrefix))
+		} else {
+			return nil
```

Comment:
> @harshavardhana Should an error be returned in this case, or is it fine to just return nil directly to bypass the request to the server?

We need to only send error if its an error, if we do know that its not an error then we are fine. 

```
var err error
if bucketPolicy == BucketPolicyNone {
        err = ErrNoSuchBucketPolicy(fmt.Sprintf("No policy exists on %s/%s", bucketName, objectPrefix))
} else {
        err = nil
}
return err
```